### DISCUSSION TITLE
Correct typing for @model_validator(mode="wrap")

### DISCUSSION THREAD

**Discussion by shufflebits** [pydantic/pydantic#10061] [Question] [answered]
pydantic==2.8.2
mypy==1.11.1

What is the correct way to type hint the following code:

```python
from pydantic import BaseModel, model_validator


class MyModel(BaseModel):

    primary: str
    secondary: str

    @model_validator(mode="wrap")
    @staticmethod
    def validate_derived(raw_values, handler):
        primary = raw_values.get("primary", "")
        if not raw_values.get("secondary"):
            raw_values["secondary"] = f"derived-from-{primary}"

        validated = handler(raw_values)
        return validated


a = MyModel(**{"primary": "foo"})
print(a.model_dump_json())
b = MyModel(**{"primary": "foo", "secondary": "bar"})
print(b.model_dump_json())


``` 

(this is just an example, - there's a proper business-logic reason for a wrap validator)

Right now, Mypy throws out:
`
main.py:9: error: Argument 1 has incompatible type "Callable[[Any, Any], Any]"; expected "ModelWrapValidator[Never] | ModelWrapValidatorWithoutInfo[Never]"  [arg-type]
`
This mode of use for wrap validators has no examples, so I'm a bit lost. There's also not much about the `handler`  callable - it seems to take an  parameter `outer_location`, but I have no idea what that does!

Changing the validator declaration to:
`    def validate_derived(raw_values: Dict[str, Any], handler: ModelWrapValidatorHandler["MyModel"]) -> "MyModel":
`
Seems not to help.

**Comment by Viicos** [comment, marked as answer]
This is how you should define your validator:

```python
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ValidatorFunctionWrapHandler, model_validator


class MyModel(BaseModel):
    primary: str
    secondary: str

    @model_validator(mode='wrap')
    @classmethod
    def validate_derived(cls, values: dict[str, Any] | MyModel, handler: ValidatorFunctionWrapHandler) -> MyModel: ...
```

Note that `values` isn't necessarily a `dict`, as it could be used as `Model.model_validate(Model.model_construct(...))`

> This mode of use for wrap validators has no examples

I agree docs should be improved. I'll try to clarify things on that side.
